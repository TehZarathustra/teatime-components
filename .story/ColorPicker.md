## Preview


### size xs

```
<div>
  some text for the baseline
  <ColorPicker
    defaultValue='FFCC99'
    name='palette'
    size='xs'/>
  {' '}
  <ColorPicker
    defaultValue='FFCC99'
    disabled={true}
    name='palette'
    size='xs'/>
</div>
```


### size s


```
<div>
  some text for the baseline
  <ColorPicker
    defaultValue='CCCCFF'
    name='palette'
    size='s'/>
  {' '}
  <ColorPicker
    defaultValue='CCCCFF'
    disabled={true}
    name='palette'
    size='s'/>
</div>
```


### size m

```
<div>
  some text for the baseline
  <ColorPicker
    name='palette'
    size='m'/>
  {' '}
  <ColorPicker
    disabled={true}
    name='palette'
    size='m'/>
</div>
```
