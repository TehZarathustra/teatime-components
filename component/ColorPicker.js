'use strict';

const {Component, PropTypes} = require('react');
const {TAB} = require('../lib/keyCode');
const {isControlled, themes} = require('../lib/tool');
const {isUndefined, map, noop, omit, prop} = require('../lib/dash');
const Overlay = require('../view/Overlay');
const React = require('react');
const Tile = require('../view/Tile');
const cc = require('classnames');

const cssModules = {
  xs: require('../style/colorPicker/colorPicker-xs.css'),
  s: require('../style/colorPicker/colorPicker-s.css'),
  m: require('../style/colorPicker/colorPicker-m.css'),
};

const omitProps = omit([
  'onChange',
  'styles',
]);

class ColorPicker extends Component {
  constructor(props) {
    super(props);

    this._controlled = isControlled(props);
    this._styles = themes(this.token);

    const value = this._controlled
      ? props.value
      : props.defaultValue;

    this.state = {
      isOpened: false,
      styles: this._styles(props),
      value: isUndefined(value) ? '' : value,
    };
  }

  componentWillReceiveProps(nextProps) {
    this._controlled = isControlled(nextProps);

    this.setState({
      styles: this._styles(nextProps),
      value: this._controlled
        ? nextProps.value
        : this.state.value,
    });
  }

  focus() {
    if (!this._input) return;
    this._input.focus();
  }

  select() {
    if (!this._input) return;
    this._input.select();
  }

  token = prop('size')

  _onChange = e => {
    const value = e.target.value;
    if (!this._controlled) this.setState({isOpened: false, value});
    this.props.onChange(e, {value});
  }

  _onClearClick = () => {
    this.setState({
      isOpened: !this.state.isOpened,
      value: '',
    });

    this.focus();
  }

  _onClick = () => {
    this.setState({isOpened: !this.state.isOpened});
  }

  _onInputFocus = () => {
    if (!this.state.isOpened) return;
    this.setState({isOpened: false});
  }

  _onKeyDown = ({keyCode}) => {
    if (this.props.disabled) return;
    if (keyCode === TAB) this.setState({isOpened: false});
  }

  _onTileClick = (e, value) => {
    if (!this._controlled) this.setState({isOpened: false, value});
    this.props.onChange(e, {value});
    this.focus();
  }

  _onOutsideClick = () => {
    if (!this.state.isOpened) return;
    this.setState({isOpened: false});
  }

  _parentNode = () => {
    return this.refs.parent;
  }

  render() {
    const {
      autoFocus,
      className,
      disabled,
      id,
      name,
      palette,
      placeholder,
      ...other,
    } = this.props;

    const {styles, value} = this.state;

    const {
      container,
      button,
      clear,
      control,
      menu,
      menuItem,
    } = styles;

    const tiles = map(color =>
      this.renderTile({
        className: menuItem,
        color,
        key: color,
        onClick: this._onTileClick,
      }), palette);

    return (
      <span
        {...omitProps(other)}
        className={cc(container, className)}
        ref='parent'>
        {this.renderButton({
          className: button,
          disabled,
          onClick: this._onClick,
          style: {
            backgroundColor: '#' + value,
          },
        })}
        {this.renderClear({
          className: clear,
          disabled,
          onClick: this._onClearClick,
          value,
        })}
        {this.renderInput({
          autoFocus,
          className: control,
          disabled,
          id,
          name,
          onChange: this._onChange,
          onFocus: this._onInputFocus,
          placeholder,
          ref: r => this._input = r,
          type: 'text',
          value,
        })}
        {this.renderMenu({
          children: tiles,
          className: menu,
          onOutsideClick: this._onOutsideClick,
          parentNode: this._parentNode,
        })}
      </span>
    );
  }

  renderButton(buttonProps) {
    return (
      <button {...buttonProps}/>
    );
  }

  renderClear(clearProps) {
    if (clearProps.disabled || !clearProps.value) return null;

    return (
      <span {...clearProps}/>
    );
  }

  renderInput(inputProps) {
    return (
      <input {...inputProps}/>
    );
  }

  renderMenu(menuProps) {
    if (!this.state.isOpened) return null;

    return (
      <Overlay {...menuProps}/>
    );
  }

  renderTile(tileProps) {
    return (
      <Tile {...tileProps}/>
    );
  }
}

ColorPicker.defaultProps = {
  onChange: noop,
  palette: [
    '000000', 'CC0000', 'CC6600',
    'CCCC00', '66CC00', '00CC00',
    '00CC66', '00CCCC', '0066CC',
    '0000CC', '6600CC', 'CC00CC',
    'CC0066', '333333', 'FF0000',
    'FF8000', 'FFFF00', '80FF00',
    '00FF00', '00FF80', '00FFFF',
    '007FFF', '0000FF', '7F00FF',
    'FF00FF', 'FF0080', '666666',
    'FF3333', 'FF9933', 'FFFF33',
    '99FF33', '33FF33', '33FF99',
    '33FFFF', '3399FF', '3333FF',
    '9933FF', 'FF33FF', 'FF3399',
    '999999', 'FF6666', 'FFB366',
    'FFFF66', 'B3FF66', '66FF66',
    '66FFB3', '66FFFF', '66B2FF',
    '6666FF', 'B266FF', 'FF66FF',
    'FF66B3', 'CCCCCC', 'FF9999',
    'FFCC99', 'FFFF99', 'CCFF99',
    '99FF99', '99FFCC', '99FFFF',
    '99CCFF', '9999FF', 'CC99FF',
    'FF99FF', 'FF99CC', 'FFFFFF',
    'FFCCCC', 'FFE6CC', 'FFFFCC',
    'E6FFCC', 'CCFFCC', 'CCFFE6',
    'CCFFFF', 'CCE5FF', 'CCCCFF',
    'E5CCFF', 'FFCCFF', 'FFCCE6',
  ],
  size: 'm',
  styles: cssModules,
};

ColorPicker.propTypes = {
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  palette: PropTypes.arrayOf(PropTypes.string),
  size: PropTypes.oneOf([
    'xs',
    's',
    'm',
  ]),
  styles: PropTypes.object,
};

module.exports = ColorPicker;
