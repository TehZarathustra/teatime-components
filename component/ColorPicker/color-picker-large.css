.colorPicker
{
  composes: colorPicker from './color-picker.css';

  font: 15px/32px arial,sans-serif;
}

.swatch
{
  composes: swatch from './color-picker.css';

  width: 32px;
  height: 32px;

  line-height: 30px;
}

.clear
{
  composes: clear from '../Input/input-large.css';
}

.control
{
  composes: control from '../Input/input-large.css';

  width: calc(100% - 31px);
}

.menu
{
  composes: menu from './color-picker.css';
}

.bottom
{
  composes: bottom from './color-picker.css';
}

.top
{
  composes: top from './color-picker.css';
}

.opened
{
  composes: opened from './color-picker.css';
}

.menuItem
{
  composes: menuItem from './color-picker.css';
}
