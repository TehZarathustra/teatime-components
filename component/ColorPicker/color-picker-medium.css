.colorPicker
{
  composes: colorPicker from './color-picker.css';

  font: 13px/28px arial,sans-serif;
}

.swatch
{
  composes: swatch from './color-picker.css';

  width: 28px;
  height: 28px;

  line-height: 26px;
}

.clear
{
  composes: clear from '../Input/input-medium.css';
}

.control
{
  composes: control from '../Input/input-medium.css';

  width: calc(100% - 27px);
}

.menu
{
  composes: menu from './color-picker.css';
}

.bottom
{
  composes: bottom from './color-picker.css';
}

.top
{
  composes: top from './color-picker.css';
}

.opened
{
  composes: opened from './color-picker.css';
}

.menuItem
{
  composes: menuItem from './color-picker.css';
}
