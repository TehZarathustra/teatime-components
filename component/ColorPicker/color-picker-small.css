.colorPicker
{
  composes: colorPicker from './color-picker.css';

  font: 13px/24px arial,sans-serif;
}

.swatch
{
  composes: swatch from './color-picker.css';

  width: 24px;
  height: 24px;

  line-height: 22px;
}

.clear
{
  composes: clear from '../Input/input-small.css';
}

.control
{
  composes: control from '../Input/input-small.css';

  width: calc(100% - 23px);
}

.menu
{
  composes: menu from './color-picker.css';
}

.bottom
{
  composes: bottom from './color-picker.css';
}

.top
{
  composes: top from './color-picker.css';
}

.opened
{
  composes: opened from './color-picker.css';
}

.menuItem
{
  composes: menuItem from './color-picker.css';
}
