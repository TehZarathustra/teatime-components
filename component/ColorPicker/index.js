/* @flow */
import {isControlled, map, noop} from '../../lib/util';
import Color from './Color';
import React, {Component} from 'react';
import classNames from 'classnames';

import colorPickerLarge from './color-picker-large.css';
import colorPickerMedium from './color-picker-medium.css';
import colorPickerSmall from './color-picker-small.css';

type Props = {
  autoFocus: boolean,
  className?: string,
  defaultValue?: string,
  direction: 'bottom' | 'top',
  disabled: boolean,
  id?: string,
  name: string,
  onBlur: (event: SyntheticEvent) => void,
  onChange: (event: SyntheticEvent, eventData: {name: string, value: string}) => void,
  onFocus: (event: SyntheticFocusEvent) => void,
  palette: Array<string>,
  placeholder?: string,
  size: 'l' | 'm' | 's',
  style?: Object,
  styles?: Object,
  value?: string,
};

type SwatchProps = {
  className: string,
  disabled?: boolean,
  onClick: Function,
  style: Object,
};

type ClearProps = {
  className: string,
  disabled?: boolean,
  onClick: Function,
} | null;

type InputProps = {
  className: string,
  disabled: boolean,
  onChange: (e: SyntheticInputEvent) => void,
  ref: Function,
};

type MenuProps = {
  children: any,
  className: string,
};

type ItemProps = {
  className: string,
  onClick: Function,
  value: string,
};

const cssModules = {
  l: colorPickerLarge,
  m: colorPickerMedium,
  s: colorPickerSmall,
};

class ColorPicker extends Component {
  props: Props

  static defaultProps: {
    autoFocus: boolean,
    direction: 'bottom' | 'top',
    disabled: boolean,
    onBlur: Function,
    onChange: Function,
    onFocus: Function,
    palette: Array<string>,
    size: 'l' | 'm' | 's',
  }

  state: {
    isMenuOpened: boolean,
    value: string,
  }

  css: {
    colorPicker: string,
    swatch: string,
    clear: string,
    control: string,
    menu: string,
    bottom: string,
    top: string,
    opened: string,
    menuItem: string,
  }

  _controlled: boolean

  _inputElement: HTMLInputElement

  constructor(props: Props) {
    super(props);

    this._controlled = isControlled(props);

    const value = this._controlled
      ? props.value
      : props.defaultValue;

    this.state = {
      isMenuOpened: false,
      value: value != null ? value : '',
    };

    this.css = props.styles || cssModules[props.size];

    (this:any)._onClearClick = this._onClearClick.bind(this);
    (this:any)._onInputBlur = this._onInputBlur.bind(this);
    (this:any)._onInputChange = this._onInputChange.bind(this);
    (this:any)._onInputFocus = this._onInputFocus.bind(this);
    (this:any)._onPaletteClick = this._onPaletteClick.bind(this);
    (this:any)._onSwatchClick = this._onSwatchClick.bind(this);
    (this:any)._saveInputRef = this._saveInputRef.bind(this);
  }

  componentWillReceiveProps(nextProps: Props) {
    this._controlled = isControlled(nextProps);

    if (this._controlled) {
      const {value} = nextProps;

      this.setState({
        value: value != null ? value : '',
      });
    }

    this.css = nextProps.styles || cssModules[nextProps.size];
  }

  focus() {
    if (this._inputElement) this._inputElement.focus();
  }

  select() {
    if (this._inputElement) this._inputElement.select();
  }

  _computeSwatchProps(): SwatchProps {
    const {disabled} = this.props;
    const {css} = this;

    return {
      className: css.swatch,
      disabled,
      onClick: this._onSwatchClick,
      style: {
        backgroundColor: '',
      },
    };
  }

  _computeClearProps(): ClearProps {
    const {disabled} = this.props;
    const {value} = this.state;
    const {css} = this;

    if (disabled || !value) return null;

    return {
      className: css.clear,
      disabled,
      onClick: this._onClearClick,
    };
  }

  _computeInputProps(): InputProps {
    const {disabled} = this.props;
    const {value} = this.state;
    const {css} = this;

    return {
      className: css.control,
      disabled,
      onBlur: this._onInputBlur,
      onChange: this._onInputChange,
      onFocus: this._onInputFocus,
      ref: this._saveInputRef,
      value,
    };
  }

  _computeMenuProps(): MenuProps {
    const {direction} = this.props;
    const {isMenuOpened} = this.state;
    const {css} = this;

    const menuItems = this._computeMenuItems(this.props.palette);

    return {
      children: menuItems,
      className: classNames(css.menu, css[direction], {
        [css.opened]: isMenuOpened,
      }),
    };
  }

  _computeMenuItems(menuItems: Array<string>) {
    if (!this.state.isMenuOpened) return null;

    const {css} = this;

    return map(color => this.renderMenuItem({
      className: css.menuItem,
      key: color,
      onClick: this._onPaletteClick,
      value: color,
    }), menuItems);
  }

  _emitChange(e: SyntheticEvent, value: string) {
    const {name, onChange} = this.props;
    if (!this._controlled) this.setState({value});
    onChange(e, {name, value});
  }

  _onClearClick(e: SyntheticEvent) {
    this._emitChange(e, '');
    this.focus();
  }

  _onInputBlur(e: SyntheticEvent) {
    this.props.onBlur(e);
  }

  _onInputChange(e: SyntheticInputEvent) {
    this._emitChange(e, e.target.value);
  }

  _onInputFocus(e: SyntheticFocusEvent) {
    this.props.onFocus(e);
  }

  _onPaletteClick(e: SyntheticEvent, color: string) {
    this.setState({isMenuOpened: false});
    this._emitChange(e, color);
  }

  _onSwatchClick(e: SyntheticEvent) {
    this.setState({isMenuOpened: !this.state.isMenuOpened});
  }

  _saveInputRef(ref: HTMLInputElement) {
    this._inputElement = ref;
  }

  render() {
    const {className, style} = this.props;
    const {css} = this;

    const swatchProps = this._computeSwatchProps();
    const clearProps = this._computeClearProps();
    const inputProps = this._computeInputProps();
    const menuProps = this._computeMenuProps();

    return (
      <span className={classNames(css.colorPicker, className)} style={style}>
        {this.renderSwatch(swatchProps)}
        {this.renderClear(clearProps)}
        {this.renderInput(inputProps)}
        {this.renderMenu(menuProps)}
      </span>
    );
  }

  renderSwatch(swatchProps: SwatchProps) {
    return (
      <button {...swatchProps}/>
    );
  }

  renderClear(clearProps: ClearProps) {
    if (!clearProps) return null;

    return (
      <span {...clearProps}/>
    );
  }

  renderInput(inputProps: InputProps) {
    return (
      <input {...inputProps}/>
    );
  }

  renderMenu(menuProps: MenuProps) {
    return (
      <div {...menuProps}/>
    );
  }

  renderMenuItem(itemProps: ItemProps) {
    return (
      <Color {...itemProps}/>
    );
  }
}

ColorPicker.defaultProps = {
  autoFocus: false,
  direction: 'bottom',
  disabled: false,
  onBlur: noop,
  onChange: noop,
  onFocus: noop,
  palette: [
    '000000', 'CC0000', 'CC6600',
    'CCCC00', '66CC00', '00CC00',
    '00CC66', '00CCCC', '0066CC',
    '0000CC', '6600CC', 'CC00CC',
    'CC0066', '333333', 'FF0000',
    'FF8000', 'FFFF00', '80FF00',
    '00FF00', '00FF80', '00FFFF',
    '007FFF', '0000FF', '7F00FF',
    'FF00FF', 'FF0080', '666666',
    'FF3333', 'FF9933', 'FFFF33',
    '99FF33', '33FF33', '33FF99',
    '33FFFF', '3399FF', '3333FF',
    '9933FF', 'FF33FF', 'FF3399',
    '999999', 'FF6666', 'FFB366',
    'FFFF66', 'B3FF66', '66FF66',
    '66FFB3', '66FFFF', '66B2FF',
    '6666FF', 'B266FF', 'FF66FF',
    'FF66B3', 'CCCCCC', 'FF9999',
    'FFCC99', 'FFFF99', 'CCFF99',
    '99FF99', '99FFCC', '99FFFF',
    '99CCFF', '9999FF', 'CC99FF',
    'FF99FF', 'FF99CC', 'FFFFFF',
    'FFCCCC', 'FFE6CC', 'FFFFCC',
    'E6FFCC', 'CCFFCC', 'CCFFE6',
    'CCFFFF', 'CCE5FF', 'CCCCFF',
    'E5CCFF', 'FFCCFF', 'FFCCE6',
  ],
  size: 'm',
};

export default ColorPicker;
