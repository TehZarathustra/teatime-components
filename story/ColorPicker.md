## Preview


### size xs

```
<div>
  some text for the baseline
  <ColorPicker
    name='palette'
    size='xs'/>
  {' '}
  <ColorPicker
    disabled={true}
    name='palette'
    size='xs'/>
</div>
```


### size s


```
<div>
  some text for the baseline
  <ColorPicker
    name='palette'
    size='s'/>
  {' '}
  <ColorPicker
    disabled={true}
    name='palette'
    size='s'/>
</div>
```


### size m

```
<div>
  some text for the baseline
  <ColorPicker
    name='palette'
    size='m'/>
  {' '}
  <ColorPicker
    disabled={true}
    name='palette'
    size='m'/>
</div>
```
