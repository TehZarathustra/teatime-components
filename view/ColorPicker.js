'use strict';

const { PropTypes } = require('react');
const { bind, hasValueProp } = require('../tool/component');
const { isUndefined, noop } = require('../tool/func');
const { isHexBased, normalizeColor, trimHash } = require('../tool/color');
const Input = require('./Input');
const OutsideEventDecorator = require('../decorator/OutsideEventDecorator');
const Overlay = OutsideEventDecorator(require('./Overlay'));
const React = require('react');
const TeatimeComponent = require('./TeatimeComponent');
const Tile = require('./Tile');
const classNames = require('classnames');
const warning = require('../tool/warning');

var didWarnForDefaultValue = false;

class ColorPicker extends TeatimeComponent {
  constructor(props) {
    super(props);

    bind(this, [
      'getParentNode',
      'onChange',
      'onInputBlur',
      'onInputFocus',
      'onKeyDown',
      'onOutsideEvent',
      'onPreviewClick',
      'onTileClick',
    ]);

    this.controlled = hasValueProp(props);

    if (process.env.NODE_ENV !== 'production' && this.controlled && !didWarnForDefaultValue) { // eslint-disable-line no-undef
      warning(typeof props.defaultValue === 'undefined',
        'ColorPicker contains an input of type text with both value and defaultValue props. ' +
        'Input elements must be either controlled or uncontrolled ' +
        '(specify either the value prop, or the defaultValue prop, but not ' +
        'both). Decide between using a controlled or uncontrolled input ' +
        'element and remove one of these props. More info: ' +
        'https://fb.me/react-controlled-components');

      didWarnForDefaultValue = true;
    }

    var value = this.controlled
      ? props.value
      : props.defaultValue;

    this.state = {
      isOpened: false,
      value: !isUndefined(value)
        ? value
        : 'FFFFFF',
    };
  }

  componentWillReceiveProps(nextProps) {
    if (this.controlled) {
      this.setState({value: nextProps.value});
    }
  }

  focus() {
    if (this.refs.control) {
      this.refs.control.focus();
    }
  }

  /**
   * @return {node|void}
   */
  getParentNode() {
    return this.refs.container;
  }

  onChange(e, data) {
    this.updateValue(e, data.value);
  }

  onInputBlur(e) {
    var nextValue = normalizeColor(this.state.value);
    if (nextValue === this.state.value) return;
    this.updateValue(e, normalizeColor(nextValue || 'FFFFFF'));
  }

  onInputFocus() {
    // @todo add possibility to control focus from outside
    if (this.state.isOpened === false) return;
    this.setState({isOpened: false});
  }

  onKeyDown(e) {
    if (this.props.disabled) return;

    const preview = this.refs.preview;

    if (e.keyCode === 9 && preview && preview === e.target) { // tab
      this.setState({isOpened: false});
    }
  }

  onPreviewClick() {
    this.setState({isOpened: !this.state.isOpened});
  }

  onOutsideEvent() {
    if (this.state.isOpened === false) return;
    this.setState({isOpened: false});
  }

  onTileClick(e, color) {
    this.updateValue(e, trimHash(color));
    this.focus();
  }

  updateValue(e, value) {
    if (!this.controlled) {
      this.setState({isOpened: false, value});
    }

    this.props.onChange(e, {isOpened: false, value});
  }

  render() {
    return (
      <div
        className={classNames(this.style('container'), {
          [this.style('isFixedWrapper')]: this.props.hasFixedWidth,
        }, this.props.className)}
        onKeyDown={this.onKeyDown}
        ref='container'>
        {this.renderPreview()}
        <Input
          {...this.knownProps()}
          className={undefined}
          defaultValue={undefined}
          onBlur={this.onInputBlur}
          onChange={this.onChange}
          onFocus={this.onInputFocus}
          ref='control'
          styles={this.styles()}
          value={this.state.value}/>
        {this.renderMenu()}
      </div>
    );
  }

  renderPreview() {
    const value = this.state.value;

    return (
      <button
        className={this.style('preview')}
        disabled={this.props.disabled}
        onClick={this.onPreviewClick}
        ref='preview'
        style={{
          background: `#${isHexBased(value) ? trimHash(value) : 'fff'}`,
        }}/>
    );
  }

  renderMenu() {
    return (
      <Overlay
        className={classNames(this.style('menu'), {
          [this.style('isClosedMenu')]: !this.state.isOpened,
          [this.style('isOpenedMenu')]: this.state.isOpened,
          [this.style('isFixedMenu')]: this.props.hasFixedWidth,
        })}
        getParentNode={this.getParentNode}
        onOutsideEvent={this.onOutsideEvent}>
        {this.renderTiles()}
      </Overlay>
    );
  }

  renderTiles() {
    if (!this.state.isOpened) {
      return null;
    }

    const palette = this.props.palette;
    const tiles = [];

    for (var length = palette.length, i = 0; i < length; ++i) {
      const tile = palette[i];

      tiles.push((
        <Tile
          color={`#${tile}`}
          key={`_${tile}`}
          onClick={this.onTileClick}
          styles={this.styles()}/>
      ));
    }

    return tiles;
  }
}

ColorPicker.defaultProps = {
  hasFixedWidth: false,
  onChange: noop,
  palette: [
    '000000',
    'CC0000',
    'CC6600',
    'CCCC00',
    '66CC00',
    '00CC00',
    '00CC66',
    '00CCCC',
    '0066CC',
    '0000CC',
    '6600CC',
    'CC00CC',
    'CC0066',
    '333333',
    'FF0000',
    'FF8000',
    'FFFF00',
    '80FF00',
    '00FF00',
    '00FF80',
    '00FFFF',
    '007FFF',
    '0000FF',
    '7F00FF',
    'FF00FF',
    'FF0080',
    '666666',
    'FF3333',
    'FF9933',
    'FFFF33',
    '99FF33',
    '33FF33',
    '33FF99',
    '33FFFF',
    '3399FF',
    '3333FF',
    '9933FF',
    'FF33FF',
    'FF3399',
    '999999',
    'FF6666',
    'FFB366',
    'FFFF66',
    'B3FF66',
    '66FF66',
    '66FFB3',
    '66FFFF',
    '66B2FF',
    '6666FF',
    'B266FF',
    'FF66FF',
    'FF66B3',
    'CCCCCC',
    'FF9999',
    'FFCC99',
    'FFFF99',
    'CCFF99',
    '99FF99',
    '99FFCC',
    '99FFFF',
    '99CCFF',
    '9999FF',
    'CC99FF',
    'FF99FF',
    'FF99CC',
    'FFFFFF',
    'FFCCCC',
    'FFE6CC',
    'FFFFCC',
    'E6FFCC',
    'CCFFCC',
    'CCFFE6',
    'CCFFFF',
    'CCE5FF',
    'CCCCFF',
    'E5CCFF',
    'FFCCFF',
    'FFCCE6',
  ],
};

// @todo add color validation for value prop
ColorPicker.propTypes = {
  hasFixedWidth: PropTypes.bool,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  palette: PropTypes.array,
  styles: PropTypes.shape({
    clear: PropTypes.string.isRequired,
    container: PropTypes.string.isRequired,
    control: PropTypes.string.isRequired,
    isClosedMenu: PropTypes.string.isRequired,
    isFixedMenu: PropTypes.string.isRequired,
    isFixedWrapper: PropTypes.string.isRequired,
    isOpenedMenu: PropTypes.string.isRequired,
    item: PropTypes.string.isRequired,
    line: PropTypes.string.isRequired,
    menu: PropTypes.string.isRequired,
    preview: PropTypes.string.isRequired,
    wrapper: PropTypes.string.isRequired,
  }),
};

ColorPicker.unwantedProps = [
  'hasFixedWidth',
  'palette',
  ...TeatimeComponent.unwantedProps,
];

module.exports = ColorPicker;
