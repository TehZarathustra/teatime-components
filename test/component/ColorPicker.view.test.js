'use strict';

import ColorPicker from '../../component/ColorPicker';
import React from 'react';
import renderer from 'react-test-renderer';

test('defaultProp #size should have a value `m`', () => {
  expect(ColorPicker.defaultProps.size).toBe('m');
});

test('Renders correctly with default props', () => {
  const tree = renderer.create(
    <ColorPicker
      name='control'/>
  );

  expect(tree).toMatchSnapshot();
});

test('Renders correctly with defined props', () => {
  const tree = renderer.create(
    <ColorPicker
      autoFocus
      className='outer'
      disabled
      id='controls-id'
      name='control'
      placeholder='rrggbb'
      value='#bebebe'/>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
