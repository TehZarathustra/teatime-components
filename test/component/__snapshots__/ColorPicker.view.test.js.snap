// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly with default props 1`] = `
<span
  className="colorPicker"
  style={undefined}
>
  <button
    className="swatch"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "",
      }
    }
  />
  <input
    className="control"
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value=""
  />
  <div
    className="menu bottom"
  />
</span>
`;

exports[`Renders correctly with defined props 1`] = `
<span
  className="colorPicker outer"
  style={undefined}
>
  <button
    className="swatch"
    disabled={true}
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "",
      }
    }
  />
  <input
    className="control"
    disabled={true}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value="#bebebe"
  />
  <div
    className="menu bottom"
  />
</span>
`;
